### PR TITLE
grass.pygrass: GridModule clean up temporary mapsets when exception occurs

### PR DESCRIPTION
**Pull request by tmszi**
**Describe the bug**
`run()` method of the `GridModule` class instance doesn't clean up temporary MAPSETs when exception occurs (input raster map doesn't exists).

**To Reproduce**

1.  Launch GRASS GIS with **nc_spm_08_grass7** LOCATION and **PERMANENT** MAPSET
2.  Launch this Python script
```python
#!/usr/bin/env python3

from grass.pygrass.modules.grid.grid import GridModule

def main():
    grd = GridModule("r.slope.aspect",
                    width=100, height=100, overlap=2,
                    processes=None, split=False, elevation="non_exist_elevation", 
                    slope="slope", aspect="aspect", overwrite=True)
    grd.run()

if __name__ == '__main__':
    main()
```
3. See error
```
GRASS nc_spm_08_grass7/PERMANENT:grass > python /tmp/test.py 
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
ERROR: Raster map <non_exist_elevation> not found
Traceback (most recent call last):
  File "/tmp/test.py", line 13, in <module>
    main()
  File "/tmp/test.py", line 10, in main
    grd.run()
  File "/usr/lib64/grass83/etc/python/grass/pygrass/modules/grid/grid.py", line 675, in run
    self.patch()
  File "/usr/lib64/grass83/etc/python/grass/pygrass/modules/grid/grid.py", line 712, in patch
    rpatch_map(
  File "/usr/lib64/grass83/etc/python/grass/pygrass/modules/grid/patch.py", line 94, in rpatch_map
    rtype.open("r")
  File "/usr/lib64/grass83/etc/python/grass/pygrass/raster/__init__.py", line 218, in open
    raise OpenError(str_err)
grass.exceptions.OpenError: The map does not exist, I can't open in 'r' mode
```

5. Check the list of LOCATION MAPSETs

```
GRASS nc_spm_08_grass7/PERMANENT:grass > tree -d -L 1 $(g.gisenv get=GISDBASE)/$(g.gisenv get=LOCATION_NAME)
/home/tomas/grassdata/nc_spm_08_grass7
├── climate_2000_2012
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_000_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_000_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_000_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_000_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_000_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_000_005
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_001_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_001_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_001_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_001_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_001_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_001_005
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_002_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_002_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_002_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_002_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_002_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_002_005
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_003_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_003_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_003_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_003_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_003_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25275_003_005
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_000_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_000_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_000_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_000_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_000_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_000_005
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_001_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_001_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_001_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_001_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_001_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_001_005
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_002_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_002_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_002_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_002_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_002_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_002_005
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_003_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_003_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_003_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_003_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_003_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_003_005
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_004_000
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_004_001
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_004_002
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_004_003
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_004_004
├── grid_r_slope_aspect_gentoo_gnu_linux_25704_004_005
├── landsat
├── PERMANENT
└── tomas
```

**Expected behavior**
Clean up the temporary MAPSETs when an exception (input raster map does not exist) occurs during the `run()` method of an instance of the `GridModule` class.

**System description (please complete the following information):**

- Operating System: all
- GRASS GIS version: all

### CODE REVIEW

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
Thanks for fixing this separately from the other issue. I think the general idea is good, but some changes would make it better - see comments.

I think a test would be doable for the specific bug being fixed. pytest seems like a good choice for that. The "positive" case where everything works is probably easier to test with grass.gunittest. There are some doctest-based tests for GridModule, but I don't know how good they are and there seems to be some confusion between tests of grass.pygrass.modules and grass.pygrass.modules.grid.

Generally, it seems that the resource handling should be more clear. The resource allocation and cleanup should be linked together in a better way, e.g., by having the cleanup set up where the resources are allocated. This may overall improve the class, but for fixing the bug, the current approach in this PR is enough.

File: python/grass/pygrass/modules/tests/grass_pygrass_grid_test.py [hunk 1]
```diff
@@ -127,16 +127,22 @@ def run_grid_module():
...
-@pytest.mark.parametrize("clean", [True, False])
-def test_cleans(tmp_path, clean):
+@pytest.mark.parametrize(
+    "clean, surface",
+    [
+        (True, False),
+        ("surface", "non_exist_surface"),
+    ],
+)
+def test_cleans(tmp_path, clean, surface):
```
Comment (wenzeslaus) [inline python/grass/pygrass/modules/tests/grass_pygrass_grid_test.py:137]: Is you intention to test all combinations? I think this is done with ([pytest-mark-parametrize](https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions)):

```python
@pytest.mark.parametrize("x", [0, 1])
@pytest.mark.parametrize("y", [2, 3])
def test_foo(x, y):
    pass
```

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
This is great - short and sweet!

I have two comments, the test actually requires some attention.

File: python/grass/pygrass/modules/tests/grass_pygrass_grid_test.py (hunk 1 above)
Comment (tmszi) [reply python/grass/pygrass/modules/tests/grass_pygrass_grid_test.py:137]: I use same pytest-mark-parametrize syntax as use `test_tilling` test function. https://github.com/OSGeo/grass/blob/8abf34e2143f2f9babbeac7b1a62db0393c09ae0/python/grass/pygrass/modules/tests/grass_pygrass_grid_test.py#L222-L230

Comment (wenzeslaus) [reply python/grass/pygrass/modules/tests/grass_pygrass_grid_test.py:137]: The syntax itself is okay. Each says something different. One `mark.parametrize` has a list of all combinations for the given parameters while multiple mark.parametrize are combined into n x m matrix.

Comment (tmszi) [reply python/grass/pygrass/modules/tests/grass_pygrass_grid_test.py:137]: Yes you right, I understand now.

File: python/grass/pygrass/modules/grid/grid.py [hunk 2]
```diff
@@ -657,7 +669,9 @@ def run(self, patch=True, clean=True):
             pool.close()
             pool.join()
             if not result.successful():
-                raise RuntimeError(_("Execution of subprocesses was not successful"))
+                raise RuntimeError(
+                    _("Execution of subprocesses was not successful"),
+                )
```
[the remark this answers is not in the document]
Comment (tmszi) [reply python/grass/pygrass/modules/grid/grid.py]: It looks like it does.

**Review by wenzeslaus — APPROVED** [review]
Great!

**Pull request merged**